force-app/main/default/classes/DMLOperations.cls

```diff
@@ -181,14 +257,55 @@ public with sharing class DMLOperations {
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
 		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
+    
+		 // Create a map to store Account Name to Account Id
+        Map<String, Account> accountMap = new Map<String, Account>();
+
+         // Collect all unique account names from the contacts
+        Set<String> accountNames = new Set<String>();
+        for (Contact contact : contactsToUpsert) {
+            accountNames.add(contact.LastName);
+        }
+    
+         // Query existing accounts with these names
+        List<Account> existingAccounts = [SELECT Id, Name FROM Account WHERE Name IN :accountNames];
+        for (Account acc : existingAccounts) {
+        accountMap.put(acc.Name, acc);
+        }
+    
+         // List to hold new accounts to be inserted
+        List<Account> accountsToInsert = new List<Account>();
+    
+         // Loop through each contact to create missing accounts and link them
+        for (Contact contact : contactsToUpsert) {
+            String accountName = contact.LastName;
+            if (!accountMap.containsKey(accountName)) {
+                 // Account does not exist, create a new one
+                Account newAccount = new Account(Name = accountName);
+                accountsToInsert.add(newAccount);
+                 accountMap.put(accountName, newAccount); // Add the new account to the map
+            }
+             // Associate the contact with the account
+            contact.AccountId = accountMap.get(accountName).Id;
+        }
+    
+         // Insert the new accounts
+        if (!accountsToInsert.isEmpty()) {
+            upsert accountsToInsert;
+             // Update the accountMap with the newly created account IDs
+            for (Account acc : accountsToInsert) {
+                accountMap.put(acc.Name, acc);
+            }
+    
+             // Update the contacts with the new Account IDs
+            for (Contact contact : contactsToUpsert) {
+                String accountName = contact.LastName;
+                contact.AccountId = accountMap.get(accountName).Id;
+            }
```

Comment:
Try using the existing upsertAccount method. Some modification may need to be make if you want to bulkify it